core/server/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -831,15 +831,24 @@ public void addInodeFromJournal(JournalEntry entry) throws AccessControlExceptio
 
       if (directory.getName().equals(ROOT_INODE_NAME)) {
         // This is the root inode. Clear all the state, and set the root.
-        if (SecurityUtils.isSecurityEnabled() && mRoot != null && !mRoot.getOwner()
-            .equals(directory.getOwner())) {
+        // For backwards-compatibility:
+        // Empty owner in journal entry indicates that previous journal has no security. In this
+        // case, the journal is allowed to be applied to the new inode with security turned on.
+        if (SecurityUtils.isSecurityEnabled() && mRoot != null && !directory.getOwner().isEmpty()
+            && !mRoot.getOwner().equals(directory.getOwner())) {
           // user is not the owner of journal root entry
           throw new AccessControlException(
               ExceptionMessage.PERMISSION_DENIED.getMessage("Unauthorized user on root"));
         }
         mInodes.clear();
         mPinnedInodeFileIds.clear();
         mRoot = directory;
+        // If journal entry has no security enabled, change the replayed inode permission to be 0777
+        // for backwards-compatibility.
+        if (SecurityUtils.isSecurityEnabled() && mRoot != null && mRoot.getOwner().isEmpty()
```

Comment:
AFAICT, it's possible for `mRoot` to be `null`:
`InodeTree#initializeRoot` is only called by `FileSystemMaster` during `start()`, while `addInodeFromJournal` (and `processJournalCheckpoint`) can be called from any `AbstractMaster` via `JournalTailer`.